pkg/mtbroker/ingress/ingress_handler.go

```diff
@@ -53,25 +55,46 @@ type Handler struct {
 	Defaulter client.EventDefaulter
 	// Reporter reports stats of status code and dispatch time
 	Reporter StatsReporter
+	// BrokerLister gets broker objects
+	BrokerLister eventinglisters.BrokerLister
 
 	Logger *zap.Logger
+}
 
-	getChannelURL func(string, string, string) url.URL
+func (h *Handler) getBroker(name, namespace string) (*eventingv1.Broker, error) {
```

Comment:
Why are we using broker as the storage mechanism instead of doing a getChannel for the duck type and using the actual status.address from it? It seems simpler and more correct and we could also be able to understand the status of the channel and display richer error message in case it goes wonky.

Reply:
Some reasons:

- **It makes the RBAC for broker ingress simpler.** Only needs read access to Brokers instead of all Channelables.
- **It reduces the number of watches and caches in a scalable resource.**
  - I may be misunderstanding how the duck tracker works, but I believe it creates a watch and informer for each concrete implementation of a duck type. Scaling the broker ingress will create N watches on the apiserver per pod where N is the number of channels installed.
  - Each ingress pod will carry an informer cache of all channels in the cluster even if they have nothing to do with brokers. It doesn't seem simpler to me to make the data plane manage those caches. The controller already has this cache and knows the specific piece of information the data plane needs.
- **The user can see the address of the channel being used by the broker.**
  - This is a minor benefit but it's there. With the data plane doing its own internal discovery of the channel address, there's no indication to the user what address is actually used.


I'm not sure how using a duck informer would be more correct than this approach. @vaikas do you see a correctness issue here?

As for channel status, seems like that would be available in the `TriggerChannelReady` condition on the Broker object.